### PR TITLE
Add for wind/unwind support to tbc

### PR DESCRIPTION
**Pull request by marcopeereboom**
**Summary**

This PR adds support for detecting canonical chain and unwinding of indexes during reorgs.

**Changes**

### CODE REVIEW

File: service/tbc/crawler.go [hunk 1]
```diff
@@ -66,6 +66,25 @@ func (s *Server) TxIndexHash(ctx context.Context) (*HashHeight, error) {
...
+// findCanonicalHash determines which hash is on the canonical chain.by walking
+// back the chain from the provided end point. It returns the index in bhs of
+// the correct hash. On failure it returns -1 DELIBERATELY to crash the caller
+// if error is not checked.
+func (s *Server) findCanonicalHash(ctx context.Context, endHash *chainhash.Hash, bhs []tbcd.BlockHeader) (int, error) {
+	switch len(bhs) {
+	case 1:
+		return 1, nil // most common fast path
+	case 0:
+		return -1, fmt.Errorf("no blockheaders provided")
```
Comment (joshuasing) [inline service/tbc/crawler.go:78]: `errors.New` can be used here, as no formatting is required.

File: database/tbcd/database.go [hunk 2]
```diff
@@ -82,6 +82,12 @@ type Database interface {
 	UtxosByScriptHash(ctx context.Context, sh ScriptHash, start uint64, count uint64) ([]Utxo, error)
 }
 
+// XXX there exist various types in this file that need to be reevaluated.
+// Such ass BlockHash, ScriptHash etc. They exist for convenience reasons but
```
Comment (joshuasing) [inline database/tbcd/database.go:86]:
```suggestion
// Such as BlockHash, ScriptHash etc. They exist for convenience reasons but
```

**Review by joshuasing — APPROVED** [review]

**Review by ClaytonNorthey92 — APPROVED** [review]

File: database/tbcd/level/level.go [hunk 3]
```diff
@@ -856,10 +877,14 @@ func (l *ldb) UtxosByScriptHash(ctx context.Context, sh tbcd.ScriptHash, start u
 	return utxos, nil
 }
 
-func (l *ldb) BlockUtxoUpdate(ctx context.Context, utxos map[tbcd.Outpoint]tbcd.CacheOutput) error {
+func (l *ldb) BlockUtxoUpdate(ctx context.Context, direction int, utxos map[tbcd.Outpoint]tbcd.CacheOutput) error {
 	log.Tracef("BlockUtxoUpdate")
 	defer log.Tracef("BlockUtxoUpdate exit")
 
+	if !(direction == 1 || direction == -1) {
```
[the remark this answers is not in the document]
Comment (marcopeereboom) [reply database/tbcd/level/level.go:864]: We have debated this endlessly and so far we haven't landed on something succinct enough.

File: service/tbc/crawler.go [hunk 4]
```diff
@@ -66,6 +69,43 @@ func (s *Server) TxIndexHash(ctx context.Context) (*HashHeight, error) {
...
+	// connecting block. There is no shortcut possible without hitting edge
+	// conditions.
+	for k, v := range bhs {
+		h := endHash
+		for {
+			bh, err := s.db.BlockHeaderByHash(ctx, h[:])
+			if err != nil {
+				return -1, fmt.Errorf("block header by hash: %w", err)
+			}
+			if h.IsEqual(v.BlockHash()) {
```
[the remark this answers is not in the document]
Comment (marcopeereboom) [reply service/tbc/crawler.go:95]: Resolved by findCanonicalChain call

File: service/tbc/crawler.go [hunk 5]
```diff
@@ -342,12 +612,67 @@ func (s *Server) UtxoIndexer(ctx context.Context, endHash *chainhash.Hash) error
...
+			Height: 0,
+		}
+	}
+
+	// XXX make sure there is no gap between start and end or vice versa.
+	startBH, err := s.db.BlockHeaderByHash(ctx, utxoHH.Hash[:])
+	if err != nil {
+		return fmt.Errorf("blockheader hash: %w", err)
+	}
+	direction, err := s.UtxoIndexIsLinear(ctx, endHash)
```
[the remark this answers is not in the document]
Comment (marcopeereboom) [reply service/tbc/crawler.go:623]: Right. We have to dedup this code. But that is a next PR thing.

File: service/tbc/crawler.go [hunk 6]
```diff
@@ -342,12 +612,67 @@ func (s *Server) UtxoIndexer(ctx context.Context, endHash *chainhash.Hash) error
...
+	// Verify exit condition hash
+	if endHash == nil {
+		return errors.New("must provide an end hash")
+	}
+	endBH, err := s.db.BlockHeaderByHash(ctx, endHash[:])
+	if err != nil {
+		return fmt.Errorf("blockheader hash: %w", err)
+	}
+
+	// Verify start point is not after the end point
```
[the remark this answers is not in the document]
Comment (marcopeereboom) [reply service/tbc/crawler.go:606]: Yes. The code that does that is right underneath.

File: database/tbcd/level/level.go [hunk 7]
```diff
@@ -762,11 +769,26 @@ func (l *ldb) SpendOutputsByTxId(ctx context.Context, txId tbcd.TxId) ([]tbcd.Sp
 	var key [1 + 32]byte
 	key[0] = 's'
 	copy(key[1:], txId[:])
+	debug := false
```
[the remark this answers is not in the document]
Comment (marcopeereboom) [reply database/tbcd/level/level.go]: Are you sure you are on latest?

File: cmd/hemictl/hemictl.go [hunk 8]
```diff
@@ -488,6 +506,24 @@ func tbcdb() error {
 			fmt.Printf("%v\n", si[k])
 		}
 
+	case "_spendoutputsbytxid":
```
[the remark this answers is not in the document]
Comment (marcopeereboom) [reply cmd/hemictl/hemictl.go]: the work is to kill `DB`. This is one of those remainers until we finish the work.

**Pull request merged**